pyomo/core/plugins/transform/scaling.py

```diff
@@ -0,0 +1,269 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright 2017 National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and 
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain 
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+from pyomo.util.plugin import alias
+from pyomo.core.base import Var, Constraint, Objective, Param, _VarData, _ConstraintData, _ObjectiveData, Suffix, value, ComponentUID
+from pyomo.core.plugins.transform.hierarchy import Transformation
+from pyomo.core.kernel import ComponentMap
+
+from pyomo.core.base import expr_coopr3
+from pyomo.core.base.expr import clone_expression
+#from pyomo.core.base import expr_common as common
+
+def _substitute_var_in_constraint(constraint, substitution_map):
+    _substitute_var(constraint._body, substitution_map)
+    
+def _substitute_var(expression, sub_component_map):
+    stack = list()
+    # this is a stack of the expression nodes to process
+    stack.append(expression)
+
+    while stack:
+        e = stack.pop()
+        e_type = type(e)
+        if e_type is expr_coopr3._ProductExpression:
+            # _ProductExpression is fundamentally different in _Coopr3
+            for i, child in enumerate(e._numerator):
+                ## TODO: don't need the isinstance... just check in sub_component_map
+                if isinstance(child, _VarData):
+                    if child in sub_component_map:
+                        e._numerator[i] = clone_expression(sub_component_map[i])
+                else:
+                    stack.append(child)
+            for i, child in enumerate(e._denominator):
+                if isinstance(child, _VarData):
+                    if child in sub_component_map:
+                        e._denominator[i] = sub_component_map[i]
+                else:
+                    stack.append(child)
+        elif hasattr(e, '_args'):
+            for i, child in enumerate(e._args):
+                if isinstance(child, _VarData):
+                    if child in sub_component_map:
+                        new_expr = clone_expression(sub_component_map[child])
+                        e._args = _replace_entry_in_list_or_tuple(e._args, new_expr, i)
+                else:
+                    stack.append(child)
+
```

Comment:
What about `Expression` objects?  Should you clone the subexpression before substituting?  It would break the "shared subexpression", but would prevent double substitutions.